Tests/GRPCTests/ZeroLengthWriteTests.swift

```diff
@@ -0,0 +1,167 @@
+import NIO
+import NIOSSL
+import NIOTransportServices
+import GRPC
+import GRPCSampleData
+import EchoModel
+import EchoImplementation
+import XCTest
+
+final class ZeroLengthWriteTests: GRPCTestCase {
+  func clientBuilder(group: EventLoopGroup, secure: Bool, debugInitializer: @escaping (Channel) -> EventLoopFuture<Void>) -> ClientConnection.Builder {
+    if secure {
+      return ClientConnection.secure(group: group)
+        .withTLS(trustRoots: .certificates([SampleCertificate.ca.certificate]))
+        .withDebugChannelInitializer(debugInitializer)
+    } else {
+      return ClientConnection.insecure(group: group)
+        .withDebugChannelInitializer(debugInitializer)
+    }
+  }
+
+  func serverBuilder(group: EventLoopGroup, secure: Bool, debugInitializer: @escaping (Channel) -> EventLoopFuture<Void>) -> Server.Builder {
+    if secure {
+      return Server.secure(
+        group: group,
+        certificateChain: [SampleCertificate.server.certificate],
+        privateKey: SamplePrivateKey.server
+      ).withTLS(trustRoots: .certificates([SampleCertificate.ca.certificate]))
+      .withDebugChannelInitializer(debugInitializer)
+    } else {
+      return Server.insecure(group: group)
+        .withDebugChannelInitializer(debugInitializer)
+    }
+  }
+
+  func makeServer(group: EventLoopGroup, secure: Bool, debugInitializer: @escaping (Channel) -> EventLoopFuture<Void>) throws -> Server {
+    return try self.serverBuilder(group: group, secure: secure, debugInitializer: debugInitializer)
+      .withServiceProviders([self.makeEchoProvider()])
+      .withLogger(self.serverLogger)
+      .bind(host: "localhost", port: 0)
+      .wait()
+  }
+
+  func makeClientConnection(group: EventLoopGroup, secure: Bool, port: Int, debugInitializer: @escaping (Channel) -> EventLoopFuture<Void>) throws -> ClientConnection {
+    return self.clientBuilder(group: group, secure: secure, debugInitializer: debugInitializer)
+      .withBackgroundActivityLogger(self.clientLogger)
+      .connect(host: "localhost", port: port)
+  }
+
+  func makeEchoProvider() -> Echo_EchoProvider { return EchoProvider() }
+
+  func makeEchoClient(group: EventLoopGroup, secure: Bool, port: Int, debugInitializer: @escaping (Channel) -> EventLoopFuture<Void>) throws -> Echo_EchoClient {
+    return Echo_EchoClient(
+      channel: try self.makeClientConnection(group: group, secure: secure, port: port, debugInitializer: debugInitializer),
+      defaultCallOptions: self.callOptionsWithLogger
+    )
+  }
+
+  func zeroLengthWriteExpectation() -> XCTestExpectation {
+    let expectation = self.expectation(description: "Expecting zero length write workaround")
+    expectation.expectedFulfillmentCount = 1
+    expectation.assertForOverFulfill = true
+    return expectation
+  }
+
+  func noZeroLengthWriteExpectation() -> XCTestExpectation {
+    let expectation = self.expectation(description: "Not expecting zero length write workaround")
+    expectation.isInverted = true
```

Comment:
Rather than inverting the expectation, can we fulfil our expectation when we see the error from not finding the handler instead?

Reply:
We can do, but it breaks the test symmetry: I have to write two test flows instead. Is that your preference?

Reply:
Why two test flows? Couldn't we just pass two `(Result<NIOFilterEmptyWritesHandler, Error>) -> Void` closures through to `_runTest` and `debugPipelineExpectation` and run them run in an `always` block rather than passing through the expectation?

Reply:
Oh, yeah, we can do that. It's a bit less clean at the test site, but works fine.